File: DP/Top_Down_fibo.cpp
#include<bits/stdc++.h>
using namespace std;

int fibo(int n,int dp[])
{
	if(n==0 or n==1)
		return n;


	if(dp[n]!=0){
		return dp[n];
	}

	int ans;
	ans = fibo(n-1,dp) + fibo(n-2,dp);

	return dp[n] = ans;

}

int main()
{
	int n = 12;
	int dp[1000] = {0};

	cout<<fibo(n,dp);
}

File: DP/Top_Down_fibo.py
def fibo(n,dp):
	# Base Case
	if n==1 or n==0:
		return n

	if(dp[n]!=0):
		return dp[n]

	dp[n] = fibo(n-1,dp) + fibo(n-2,dp)
	return dp[n]


def main(): 
	n = 8
	dp = [0]*(n+1)
	ans = fibo(n,dp)
	print(ans)

main()